### PR TITLE
Improve python console help

### PR DESCRIPTION
**Pull request by YoannQDQ**
## Description

Add a special '?' command that displays the following text:

```
QGIS Python Console
======================================

The console is a Python interpreter that allows you to execute python commands.
Modules from QGIS (analysis, core, gui, 3d) and Qt (QtCore, QtGui, QtNetwork, 
QtWidgets, QtXml) as well as Python's math, os, re and sys modules are already
imported and can be used directly.

Useful variables:

- iface.mainWindow() will return the Qt Main Window
- iface.mapCanvas() will return the map canvas
- iface.layerTreeView() will return the Layer Tree
- iface.activeLayer() will return the active layer
- QgsProject.instance() will return the current project

From the console, you can type the following special commands:

    - _pyqgis, _pyqgis(object): Open the QGIS Python API (or the Qt documentation) in a web browser
    - _api, _api(object): Open the QGIS C++ API (or the Qt documentation) in a web browser
    - _cookbook: Open the PyQGIS Developer Cookbook in a web browser
    - System commands: Any command starting with an exclamation mark (!) will be executed by the system shell. Examples:
        !gdalinfo --formats: List all available GDAL drivers
        !ogr2ogr --help: Show help for the ogr2ogr command
        !ping www.qgis.org: Ping the QGIS website
        !pip install black: install black python formatter using pip (if available)
    - ?: Show this help
```

--- 

_pyqgis and _api can now take a parameter (instance, or class), and will display the matching documentation page in (Py)QGIS doc or Qt doc.

doc

### CODE REVIEW

**Comment by nyalldawson** [comment]
Nice! Much welcome changes as always :+1: 

Any chance we could port these changes to the c++ QgsCodeEditorPython class?

**Comment by YoannQDQ** [comment]
Hum. I'm not sure how it would make sense outside the pseudo python interpreter (which is currently the only `CommandInput` `QgsCodeEditor`in the whole code base).

What exactly did you think was worth porting to `QgsCodeEditorPython`?

**Comment by agiudiceandrea** [comment]
> - System commands: Any command starting with an exclamation mark (!) will be executed by the system shell. Examples:
>         !ls -l: list files in the current directory
>         !pwd: print the current directory

I suppose those commands only work on Linux and perhaps on macOS, but not on Windows. Maybe it would be useful to specify it and add an example working on Windows.

File: python/console/console_output.py [hunk 1]
```diff
@@ -85,6 +81,34 @@ def isatty(self):
...
+- iface.mapCanvas() will return the map canvas.
+- iface.layerTreeView() will return the Layer Tree.
+- iface.activeLayer() will return the active layer
+- QgsProjet.instance() will return the current project
+
+From the console, you can type the following special commands:
+
+    - _pyqgis, _pyqgis(object): Open the QGIS Python API (or the Qt documentation) in a web browser
+    - _api, _api(object): Open the QGIS C++ API (or the Qt documentation) in a web browser
+    - _cookbook: Open the PyQGIS Developer Cookbook  in a web browser
```
Comment (nicogodet) [inline python/console/console_output.py:102]:
```suggestion
    - _cookbook: Open the PyQGIS Developer Cookbook in a web browser
```

File: python/console/console_output.py [hunk 2]
```diff
@@ -85,6 +81,34 @@ def isatty(self):
...
 
+FULL_HELP_TEXT = QCoreApplication.translate("PythonConsole", """QGIS Python Console
+======================================
+
+The console is a Python interpreter that allows you to execute python commands
+QGIS libraries are preloaded and can be used directly from the console.
+
+Useful variables:
+
+- iface.mainWindow() will return the Qt Main Window
```
Comment (nicogodet) [inline python/console/console_output.py:92]:
```suggestion
- iface.mainWindow() will return the Qt Main Window.
```

File: python/console/console_output.py [hunk 3]
```diff
@@ -85,6 +81,34 @@ def isatty(self):
...
+- iface.mapCanvas() will return the map canvas.
+- iface.layerTreeView() will return the Layer Tree.
+- iface.activeLayer() will return the active layer
```
Comment (nicogodet) [inline python/console/console_output.py:95]:
```suggestion
- iface.activeLayer() will return the active layer.
```

File: python/console/console_output.py [hunk 4]
```diff
@@ -85,6 +81,34 @@ def isatty(self):
...
+- QgsProjet.instance() will return the current project
```
Comment (nicogodet) [inline python/console/console_output.py:96]:
```suggestion
- QgsProjet.instance() will return the current project.
```

File: python/console/console_output.py [hunk 5]
```diff
@@ -85,6 +81,34 @@ def isatty(self):
         return False
 
 
+FULL_HELP_TEXT = QCoreApplication.translate("PythonConsole", """QGIS Python Console
+======================================
+
+The console is a Python interpreter that allows you to execute python commands
+QGIS libraries are preloaded and can be used directly from the console.
```
Comment (kannes) [inline python/console/console_output.py:88]: Maybe like this to make it more clear?

```suggestion
QGIS and Qt modules are already imported and can be used directly.
```

or even
```
Modules from QGIS (analysis, core, gui, 3d) and Qt (QtCore, QtGui, QtNetwork, 
QtWidgets, QtXml) as well as Python's math, os, re and sys modules are already
imported and can be used directly.
```

File: python/console/console_output.py (hunk 4 above)
Comment (kannes) [reply python/console/console_output.py:96]: QgsProje**c**t with a c! :D

**Review by kannes — COMMENTED** [review]
Another great improvement, thank you! I had tried to build something similar for _pyqgis some months ago but gave up too early. This will be incredibly useful :)

File: python/console/console_output.py (hunk 1 above)
Comment (YoannQDQ) [reply python/console/console_output.py:102]: Ok

File: python/console/console_output.py (hunk 2 above)
Comment (YoannQDQ) [reply python/console/console_output.py:92]: I'd rather remove the dot from all entries

File: python/console/console_output.py (hunk 4 above)
Comment (YoannQDQ) [reply python/console/console_output.py:96]: Woops

File: python/console/console_output.py (hunk 2 above)
Comment (kannes) [reply python/console/console_output.py:94]: :+1: for no dots from me

File: python/console/console_output.py [hunk 6]
```diff
@@ -85,6 +81,36 @@ def isatty(self):
...
+======================================
+
+The console is a Python interpreter that allows you to execute python commands.
+Modules from QGIS (analysis, core, gui, 3d) and Qt (QtCore, QtGui, QtNetwork,
+QtWidgets, QtXml) as well as Python's math, os, re and sys modules are already
+imported and can be used directly.
+
+Useful variables:
+
+- iface.mainWindow() will return the Qt Main Window
```
Comment (Gustry) [inline python/console/console_output.py:94]:
```suggestion
- iface will return the current QGIS interface, class 'QgisInterface'
- iface.mainWindow() will return the Qt Main Window
```

I guess it's handy to have `iface` which leads to all variables, and also the class name
Feel free to rephrase.

File: python/console/console_output.py [hunk 7]
```diff
@@ -85,6 +81,36 @@ def isatty(self):
...
+- iface.mapCanvas() will return the map canvas
+- iface.layerTreeView() will return the Layer Tree
```
Comment (Gustry) [inline python/console/console_output.py:96]:
```suggestion
- iface.layerTreeView() will return the layer tree
```

**Review by Gustry — COMMENTED** [review]
Nice addition !
Thanks

**Comment by nyalldawson** [comment]
@YoannQDQ 
> What exactly did you think was worth porting to QgsCodeEditorPython?

Honestly all of it :) My long term goal is that there won't be anymore code in python/console, but that everything will be implemented in re-usable classes in GUI. 

My thinking is that we've got a proliferation of different languages that users can script in, or run interactive consoles in. Ie python, SQL, QGIS expressions, and likely R in future. So we can either have multiple consoles: "Python Console", "SQL Console", "QGIS Expression Console" all exposes as their own separate docks,.... OR one unified scripting console dock where you open different interpretors as tabs. I.e. from the unified dock you pick "New SQL Console" and it opens a new console for that language in a tab (and similar for the script editors).

I'll merge this so it doesn't miss today's freeze, but at some stage in the future I'd love it to be moved to the base class to help facilitate these longer term plans :+1:

**Pull request merged**